airflow/hooks/oracle_hook.py

```diff
@@ -50,20 +50,61 @@ def get_conn(self):
         as in ``{ "dsn":"some.host.address" , "service_name":"some.service.name" }``
         """
         conn = self.get_connection(self.oracle_conn_id)
+        conn_config = {
+            'user': conn.login,
+            'password': conn.password or ''
+        }
         dsn = conn.extra_dejson.get('dsn', None)
         sid = conn.extra_dejson.get('sid', None)
         mod = conn.extra_dejson.get('module', None)
 
         service_name = conn.extra_dejson.get('service_name', None)
         if dsn and sid and not service_name:
-            dsn = cx_Oracle.makedsn(dsn, conn.port, sid)
-            conn = cx_Oracle.connect(conn.login, conn.password, dsn=dsn)
+            conn_config['dsn'] = cx_Oracle.makedsn(dsn, conn.port, sid)
         elif dsn and service_name and not sid:
-            dsn = cx_Oracle.makedsn(dsn, conn.port, service_name=service_name)
-            conn = cx_Oracle.connect(conn.login, conn.password, dsn=dsn)
+            conn_config['dsn'] = cx_Oracle.makedsn(dsn, conn.port, service_name=service_name)
         else:
-            conn = cx_Oracle.connect(conn.login, conn.password, conn.host)
+            conn_config['dsn'] = conn.host
 
+        if conn.extra_dejson.get('encoding', None):
+            conn_config['encoding'] = conn.extra_dejson.get('encoding')
+            # if `encoding` is specific but `nencoding` is not
+            # `nencoding` should use same values as `encoding` to set encoding, inspired by
+            # https://github.com/oracle/python-cx_Oracle/issues/157#issuecomment-371877993
+            if not conn.extra_dejson.get('nencoding', None):
+                conn_config['nencoding'] = conn.extra_dejson.get('encoding')
+        if conn.extra_dejson.get('nencoding', None):
+            conn_config['nencoding'] = conn.extra_dejson.get('nencoding')
+        mode = conn.extra_dejson.get('mode', None)
```

Comment:
```suggestion
        mode = conn.extra_dejson.get('mode', '').lower()
```

And then we don't need `lower()` on all of the branches, or the extra `if mode:` on line 79

Reply:
That great